enterprise/server/remote_execution/dirtools/dirtools.go

```diff
@@ -357,6 +359,32 @@ func UploadTree(ctx context.Context, env environment.Env, dirHelper *DirHelper,
 				if err != nil {
 					return nil, err
 				}
+
+				ospb := &repb.OutputSymlink{
+					Path:   trimPathPrefix(fqfn, rootDir),
+					Target: target,
+				}
+				// In case the target is a dangling symlink (the target does not exist, or inaccessible),
+				// there is no way to tell whether it's a directory or a file symlink.
+				// Dangling symlinks could be used in packaging with targets that only exist on the
+				// deployment environments.
+				//
+				// Set both OutputFileSymlinks and OutputDirectorySymlinks only when the symlinks
+				// could be resolved.
+				if targetInfo, err := os.Stat(fqfn); err == nil {
+					// OutputFileSymlinks and OutputDirectorySymlinks are deprecated in REAPI v2.2,
+					// but Bazel still uses them as of v6.2.0.
+					//
+					// Set both the deprecated and the new fields to keep backward compatibility
+					// with older REAPI clients.
+					if targetInfo.IsDir() {
+						actionResult.OutputDirectorySymlinks = append(actionResult.OutputDirectorySymlinks, ospb)
+					} else {
+						actionResult.OutputFileSymlinks = append(actionResult.OutputFileSymlinks, ospb)
+					}
```

Comment:
The documentation of `output_directory_symlinks` says:

```
  // The output directories of the action that are symbolic links to other
  // directories.
```

similarly for `output_file_symlinks`:

```
  // The output files of the action that are symbolic links to other files.
```

I am interpreting this to mean that "the output files of the action" means `Action.command.output_files` and "the output directories of the action" means `Action.command.output_directories`.

If that's correct then under that interpretation, we need to check that this symlink file is exactly referenced in output_files or output_directories before we add them to the ActionResult.output_file_symlinks or ActionResult.output_directory_symlinks (respectively).

(Note that this function begins directory traversal starting from the workspace root, not any of the specific output directories or output file/symlink paths, otherwise we wouldn't need an explicit check.)